azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt

```diff
@@ -23,62 +23,101 @@ class MultiOracleScenarioRunner<
         }
     }
 
-    private val actionGenerators by lazy {
-        testInstance.runTask(TaskType.CreateActionGenerators, scenario) { implementation ->
-            val systemFactory = implementation.createSystemFactory()
-            scenario.actionGenerations(systemFactory.actionGeneratorFactory)
-        }.result!!
-    }
-
 
     fun run(): Result<AF, CF, QF, AGF> {
         var remainingOracles = oracleInstances
         val oracleResults = mutableListOf<OracleResult<AF, CF, QF, AGF>>()
+        var verificationContext: VerificationContext<AF, CF, QF, AGF>? = null
         while (oracleResults.lastOrNull()?.verified != true && remainingOracles.isNotEmpty()) {
             val oracle = remainingOracles.first()
-            remainingOracles = remainingOracles.drop(1)
-            Log.info("Attempting to verify scenario, oracle=$oracle")
-            val oracleResult = verifyWithOracle(oracle)
+            val resultBuilder = OracleResult.Builder(oracle)
+            verificationContext = establishVerificationContext(oracle, resultBuilder, verificationContext)
+            val oracleResult = if (verificationContext == null) {
+                resultBuilder.build()
+            } else {
+                Log.info("Attempting to verify scenario, oracle=$oracle")
+                verifyWithOracle(oracle, resultBuilder, verificationContext)
```

Comment:
Shouldn't the logging instruction be inside `verifyWithOracle`?

Reply:
I'd say generally you want the logging to be as high 'up' the stack as sensible so that you can reuse the function without having to inherit the logging. 

Here, specifically, the log statement is marking progress of the overall run process. It wouldn't make much difference to have it in the function, but I certainly wouldn't consider it better.